tactics/tacticals.ml

```diff
@@ -727,6 +727,32 @@ module New = struct
       let (loc,_) = evi.Evd.evar_source in
       Pretype_errors.error_unsolvable_implicit ?loc env sigma evk None
 
+  let tclMAPDELAYEDWITHHOLES accept_unresolved_holes l tac =
+    let rec aux = function
+      | [] -> tclUNIT ()
+      | x :: l ->
+        Proofview.Goal.enter begin fun gl ->
```

Comment:
Just to be sure, there are two possible places where the enter can be performed. The current code recursively enters the goal at each recursive step of `aux`, but it is also possible to hoist it out and only enter once at toplevel. Is the current behaviour the desired one? The difference is observable if e.g. `tac` generates several subgoals.

Reply:
I was needing the sigma and the env at the current stage of the proof and I didn't know how to do it without entering (`tclENV` does not give the environment of the current goal for instance). So, it is unclear to me how it would look like with a single enter at toplevel. If it is easy to explain, I'd be happy to see how it works.

Reply:
Just call the enter once, and keep the same env in the recursive call to `aux`. But it's probably not very important, I don't mind the PR being merged as-is.

Reply:
Is it the same env? with arbitrary tac that seems unlikely

Reply:
Not necessarily, but this is part of the question of the semantics. If the tactic argument starts performing non-trivial effects you can get weird behaviours with this combinator.

Reply:
I added a comment clarifying this choice, hoping it is correct for you.